Improve implementation of respect_retry_after_header

Eronarn opened the pull request:
`respect_retry_after_header` is accepted as an argument to `Retry`, but it isn't propagated on `Retry.new()`. This means that after the first retry is incremented, this setting will be lost.

Additionally, this setting isn't properly applied in cases where the `retry_after` header is present, but the intent is _not_ to respect it. This edge case can occur if you explicitly allow retries for the same codes as `RETRY_AFTER_STATUS_CODES = frozenset([413, 429, 503])`, but _don't_ respect the value in the header, and instead fall back to your `Retry`'s exponential backoff settings. In our case, we're dealing with an API that returns a large static value in this header even though we can actually try again much sooner because rate limits are governed by a token bucket.

I've attached a patch that addresses these two cases and adds a test case for one of them. However, I can't get the test suite to run locally (at all), I'm afraid.

Eronarn commented:
I took a glance at the test failures, looks spurious?

pquentin commented:
Right, they're not related to your changes and happen in other pull requests too.

sethmlarson commented:
Cycling to get an automatic rebase for CI

Eronarn commented:
Looks like it passed this time. Is this mergeable? Do I need any additional tests, contributor agreement, signature in blood, etc.? :)

sethmlarson commented on src/urllib3/util/retry.py:
```diff
@@ -192,7 +192,8 @@ def new(self, **kw):
             history=self.history,
-            remove_headers_on_redirect=self.remove_headers_on_redirect
+            remove_headers_on_redirect=self.remove_headers_on_redirect,
+            respect_retry_after_header=self.respect_retry_after_header
```
Nice catch here!

sethmlarson commented on src/urllib3/util/retry.py:
```diff
@@ -274,7 +275,7 @@ def sleep(self, response=None):
         """
 
-        if response:
+        if self.respect_retry_after_header and response:
```
Could we add a new unit test that hits this condition?

Eronarn commented:
@sethmlarson I did this:

- rebased with current master (post-black)
- moved the unit tests for this header from `util.py` to `retry.py` to centralize them
- added an additional unit test
- applied black to the modified files

What's the appropriate way for a changelog bump? Does this become a new tagged release by itself, or is there a "on master but not released" place to put changelogs?

sethmlarson commented on test/test_retry.py:
```diff
@@ -271,3 +274,47 @@ def test_retry_set_remove_headers_on_redirect(self):
+        with pytest.raises(InvalidHeader):
+            retry.parse_retry_after(value)
+
+    @pytest.mark.parametrize(
```
Can we add a test case for an http-date formatted Retry-After value? May require mocking time.time() to get a consistent result.

Eronarn replied:
Added four cases here (whether the time is in the future or not X whether we're respecting the header or not).

Eronarn commented:
@sethmlarson added the above, how's that look?

sethmlarson commented on test/test_retry.py:
```diff
@@ -271,3 +276,68 @@ def test_retry_set_remove_headers_on_redirect(self):
+            )
+        )
+
+        with mock.patch("time.sleep") as sleep_mock, mock.patch(
```
TIL about this construction! Thanks for showing me this. :)

sethmlarson commented:
This PR is only failing for Python 3.8 due to reasons not related to this PR. Once I resolve those issues I'll cycle CI and merge. Thank you so much for this! :) If you've got more time and want to help out more there are plenty of issues marked "[Contributor Friendly](https://github.com/urllib3/urllib3/issues?q=is%3Aissue+is%3Aopen+label%3A%22Contributor+Friendly+%E2%99%A5%22)" that need eyes on them.

sethmlarson commented:
Thanks for this! :D
